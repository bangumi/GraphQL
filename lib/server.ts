import * as fs from 'node:fs';
import * as path from 'node:path';

import type { FastifyInstance, FastifyRequest, FastifyServerOptions } from 'fastify';
import { fastify } from 'fastify';
import mercurius from 'mercurius';
import AltairFastify from 'altair-fastify-plugin';
import swagger from '@fastify/swagger';
import type { OpenAPIV3 } from 'openapi-types';

import { schema } from './graphql/schema';
import type { Context } from './graphql/context';
import prisma from './prisma';
import * as auth from './auth';
import * as rest from './rest';
import { pkg, projectRoot } from './config';
import { Security } from './openapi';

export async function createServer(opts: FastifyServerOptions = {}): Promise<FastifyInstance> {
  const server = fastify(opts);

  await server.register(mercurius, {
    schema,
    path: '/v0/graphql',
    graphiql: false,
    allowBatchedQueries: true,
    context: async (request: FastifyRequest): Promise<Context> => {
      return {
        prisma,
        auth: await auth.byHeader(request.headers.authorization),
      };
    },
  });

  await server.register(AltairFastify, {
    path: '/v0/altair/',
    baseURL: '/v0/altair/',
    endpointURL: '/v0/graphql',
    initialSettings: {
      theme: 'dark',
      'plugin.list': ['altair-graphql-plugin-graphql-explorer'],
    },
  });

  const swaggerUI = fs.readFileSync(path.join(projectRoot, './lib/swagger.html'));

  server.get('/v0.5/', (_, res) => {
    void res.type('text/html').send(swaggerUI);
  });

  server.get('/v0.5/openapi.json', () => {
    return server.swagger();
  });

  const openapi: Partial<OpenAPIV3.Document> = {
    info: {
      version: pkg.version,
      title: 'hello',
    },
    components: {
      securitySchemes: {
        [Security.OptionalHTTPBearer]: {
          type: 'http',
          description:
            '不强制要求用户认证，但是可能看不到某些敏感内容内容（如 NSFW 或者仅用户自己可见的收藏）',
          scheme: 'Bearer',
        },
        [Security.HTTPBearer]: {
          type: 'http',
          description: '需要使用 access token 进行认证',
          scheme: 'Bearer',
        },
      },
    },
  };

  await server.register(swagger, {
    openapi,
  });

  await server.register(rest.setup, { prefix: '/v0.5' });

  return server;
}
