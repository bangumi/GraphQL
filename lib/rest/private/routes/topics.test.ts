import { describe, expect, test } from 'vitest';

import { createServer } from '../../../server';

describe('group topics', () => {
  test('should failed on not found group', async () => {
    const app = await createServer();

    const res = await app.inject({
      url: '/p1/groups/non-existing/topics',
    });

    expect(res.statusCode).toBe(404);
    expect(res.json()).toMatchSnapshot();
  });

  test('should return data', async () => {
    const app = await createServer();

    const res = await app.inject({
      url: '/p1/groups/sandbox/topics',
    });

    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });

  test('should fetch group profile', async () => {
    const app = await createServer();

    const res = await app.inject('/p1/groups/sandbox/profile');
    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });
});
