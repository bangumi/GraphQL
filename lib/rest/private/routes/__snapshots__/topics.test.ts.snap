// Vitest Snapshot v1

exports[`group topics > should failed on not found group 1`] = `
{
  "code": "NOT_FOUND",
  "error": "Not Found",
  "message": "group not found",
  "statusCode": 404,
}
`;

exports[`group topics > should fetch group profile 1`] = `
{
  "group": {
    "createdAt": 1531631310,
    "icon": "https://lain.bgm.tv/pic/icon/s/000/00/42/4215.jpg?r=1531631345",
    "id": 4215,
    "name": "sandbox",
    "nsfw": false,
    "summary": "[s]非[/s]官方沙盒",
    "title": "沙盒",
  },
  "inGroup": false,
  "recentAddedMembers": [
    {
      "avatar": {
        "large": "https://lain.bgm.tv/pic/user/l/000/38/29/382951.jpg?r=1571167246",
        "medium": "https://lain.bgm.tv/pic/user/m/000/38/29/382951.jpg?r=1571167246",
        "small": "https://lain.bgm.tv/pic/user/s/000/38/29/382951.jpg?r=1571167246",
      },
      "id": 382951,
      "nickname": "树洞酱",
      "username": "382951",
    },
    {
      "avatar": {
        "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
        "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
        "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
      },
      "id": 427613,
      "nickname": "nickname 427613",
      "username": "427613",
    },
    {
      "avatar": {
        "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
        "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
        "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
      },
      "id": 287622,
      "nickname": "nickname 287622",
      "username": "287622",
    },
  ],
  "topics": [
    {
      "creator": {
        "avatar": {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 287622,
        "nickname": "nickname 287622",
        "username": "287622",
      },
      "id": 371602,
      "lastRepliedAt": "2022-07-15T11:47:28.000Z",
      "parentID": 4215,
    },
  ],
}
`;

exports[`group topics > should fetch topic details 1`] = `
{
  "group": {
    "createdAt": 1531631310,
    "icon": "https://lain.bgm.tv/pic/icon/s/000/00/42/4215.jpg?r=1531631345",
    "id": 4215,
    "name": "sandbox",
    "nsfw": false,
    "summary": "[s]非[/s]官方沙盒",
    "title": "沙盒",
  },
  "replies": [
    {
      "createdAt": 1658030918,
      "creator": {
        "avatar": {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 287622,
        "isFriend": false,
        "nickname": "nickname 287622",
        "username": "287622",
      },
      "id": 2056216,
      "replies": [],
      "state": 6,
      "text": "",
    },
    {
      "createdAt": 1662276469,
      "creator": {
        "avatar": {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 287622,
        "isFriend": false,
        "nickname": "nickname 287622",
        "username": "287622",
      },
      "id": 2092004,
      "replies": [],
      "state": 7,
      "text": "",
    },
    {
      "createdAt": 1662282514,
      "creator": {
        "avatar": {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 1,
        "isFriend": false,
        "nickname": "nickname 1",
        "username": "1",
      },
      "id": 2092065,
      "replies": [],
      "state": 5,
      "text": "",
    },
    {
      "createdAt": 1662282518,
      "creator": {
        "avatar": {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 1,
        "isFriend": false,
        "nickname": "nickname 1",
        "username": "1",
      },
      "id": 2092066,
      "replies": [],
      "state": 1,
      "text": "",
    },
    {
      "createdAt": 1662282523,
      "creator": {
        "avatar": {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 1,
        "isFriend": false,
        "nickname": "nickname 1",
        "username": "1",
      },
      "id": 2092067,
      "replies": [],
      "state": 2,
      "text": "",
    },
    {
      "createdAt": 1662282554,
      "creator": {
        "avatar": {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 287622,
        "isFriend": false,
        "nickname": "nickname 287622",
        "username": "287622",
      },
      "id": 2092069,
      "replies": [
        {
          "createdAt": 1662283112,
          "creator": {
            "avatar": {
              "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
              "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
              "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
            },
            "id": 287622,
            "isFriend": false,
            "nickname": "nickname 287622",
            "username": "287622",
          },
          "id": 2092074,
          "state": 0,
          "text": "sub",
        },
      ],
      "state": 0,
      "text": "q",
    },
  ],
}
`;

exports[`group topics > should return data 1`] = `
{
  "data": [
    {
      "creator": {
        "avatar": {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "id": 287622,
        "nickname": "nickname 287622",
        "username": "287622",
      },
      "id": 371602,
      "lastRepliedAt": "2022-07-15T11:47:28.000Z",
      "parentID": 4215,
    },
  ],
  "total": 1,
}
`;
